# Creating new variables

In [15]:
import numpy as np
import pandas as pd
from pandas import DataFrame, Series

In [16]:
kobe = pd.DataFrame.from_csv("./data/kobe_searching_for_new_variables.csv", sep=",")

## Creating variables crossing 'action_type' with shot zone and distance variables ('shot_zone_area', 'shot_zone_basic', 'shot_zone_range', and 'shot_distance')

Analyzing the data on Tableau we have observed that these are the only variables that exhibit remarkable variations from the average shot accuracy. 
It is possible to find the charts of the accuracy percentage according to the crossing of these categories on Tableau book 'kobe_searching_new_variables.twb' in the directory 'visualization'.

In [3]:
kobe["at_sza"] = kobe["action_type"] + "-" + kobe["shot_zone_area"]

In [4]:
kobe["at_sza"].unique()

array(['Jump Shot-Right Side(R)', 'Jump Shot-Left Side(L)',
       'Jump Shot-Left Side Center(LC)', 'Jump Shot-Right Side Center(RC)',
       'Driving Dunk Shot-Center(C)', 'Layup Shot-Center(C)',
       'Jump Shot-Center(C)', 'Running Jump Shot-Center(C)',
       'Running Jump Shot-Left Side(L)', 'Driving Layup Shot-Center(C)',
       'Running Jump Shot-Left Side Center(LC)',
       'Running Jump Shot-Right Side(R)', 'Reverse Layup Shot-Center(C)',
       'Reverse Dunk Shot-Center(C)', 'Slam Dunk Shot-Center(C)',
       'Turnaround Jump Shot-Left Side(L)',
       'Turnaround Jump Shot-Right Side(R)', 'Tip Shot-Center(C)',
       'Running Hook Shot-Center(C)', 'Alley Oop Dunk Shot-Center(C)',
       'Running Jump Shot-Right Side Center(RC)', 'Dunk Shot-Center(C)',
       'Jump Shot-Back Court(BC)', 'Turnaround Jump Shot-Center(C)',
       'Alley Oop Layup shot-Center(C)', 'Running Dunk Shot-Center(C)',
       'Driving Finger Roll Shot-Center(C)',
       'Running Layup Shot-Center(C)',

In [5]:
len(kobe["at_sza"].unique())

122

In [6]:
kobe["at_szb"] = kobe["action_type"] + "-" + kobe["shot_zone_basic"]

In [7]:
kobe["at_szb"].unique()

array(['Jump Shot-Mid-Range', 'Driving Dunk Shot-Restricted Area',
       'Layup Shot-Restricted Area', 'Jump Shot-Restricted Area',
       'Jump Shot-In The Paint (Non-RA)',
       'Running Jump Shot-In The Paint (Non-RA)',
       'Jump Shot-Above the Break 3', 'Driving Layup Shot-Restricted Area',
       'Running Jump Shot-Mid-Range', 'Reverse Layup Shot-Restricted Area',
       'Reverse Dunk Shot-Restricted Area',
       'Slam Dunk Shot-Restricted Area',
       'Layup Shot-In The Paint (Non-RA)',
       'Driving Layup Shot-In The Paint (Non-RA)',
       'Jump Shot-Right Corner 3', 'Turnaround Jump Shot-Mid-Range',
       'Tip Shot-Restricted Area', 'Running Hook Shot-Restricted Area',
       'Alley Oop Dunk Shot-Restricted Area', 'Dunk Shot-Restricted Area',
       'Jump Shot-Backcourt', 'Jump Shot-Left Corner 3',
       'Turnaround Jump Shot-In The Paint (Non-RA)',
       'Alley Oop Layup shot-In The Paint (Non-RA)',
       'Running Dunk Shot-Restricted Area',
       'Running Jump 

In [8]:
len(kobe["at_szb"].unique())

130

In [9]:
kobe["at_szr"] = kobe["action_type"] + "-" + kobe["shot_zone_range"]

In [10]:
kobe["at_szr"].unique()

array(['Jump Shot-16-24 ft.', 'Jump Shot-8-16 ft.',
       'Driving Dunk Shot-Less Than 8 ft.', 'Layup Shot-Less Than 8 ft.',
       'Jump Shot-Less Than 8 ft.', 'Running Jump Shot-8-16 ft.',
       'Jump Shot-24+ ft.', 'Driving Layup Shot-Less Than 8 ft.',
       'Running Jump Shot-16-24 ft.', 'Running Jump Shot-Less Than 8 ft.',
       'Reverse Layup Shot-Less Than 8 ft.',
       'Reverse Dunk Shot-Less Than 8 ft.',
       'Slam Dunk Shot-Less Than 8 ft.', 'Turnaround Jump Shot-8-16 ft.',
       'Turnaround Jump Shot-16-24 ft.', 'Tip Shot-Less Than 8 ft.',
       'Running Hook Shot-Less Than 8 ft.',
       'Alley Oop Dunk Shot-Less Than 8 ft.', 'Dunk Shot-Less Than 8 ft.',
       'Jump Shot-Back Court Shot', 'Alley Oop Layup shot-Less Than 8 ft.',
       'Running Dunk Shot-Less Than 8 ft.',
       'Driving Finger Roll Shot-Less Than 8 ft.',
       'Turnaround Jump Shot-Less Than 8 ft.',
       'Running Layup Shot-Less Than 8 ft.',
       'Finger Roll Shot-Less Than 8 ft.', 'Running H

In [11]:
len(kobe["at_szr"].unique())

105

In [12]:
kobe["at_sd"] = kobe["action_type"] + "-" + kobe["shot_distance"].astype(str)

In [13]:
kobe["at_sd"].unique()

array(['Jump Shot-18', 'Jump Shot-15', 'Jump Shot-16', 'Jump Shot-22',
       'Driving Dunk Shot-0', 'Jump Shot-14', 'Layup Shot-0',
       'Jump Shot-2', 'Jump Shot-12', 'Running Jump Shot-12',
       'Jump Shot-25', 'Jump Shot-17', 'Jump Shot-9', 'Jump Shot-5',
       'Jump Shot-20', 'Driving Layup Shot-0', 'Running Jump Shot-11',
       'Running Jump Shot-18', 'Running Jump Shot-7', 'Jump Shot-26',
       'Jump Shot-19', 'Jump Shot-8', 'Jump Shot-13', 'Layup Shot-1',
       'Reverse Layup Shot-0', 'Jump Shot-27', 'Layup Shot-2',
       'Reverse Dunk Shot-0', 'Slam Dunk Shot-0', 'Jump Shot-7',
       'Jump Shot-28', 'Jump Shot-24', 'Jump Shot-11', 'Layup Shot-5',
       'Driving Layup Shot-5', 'Jump Shot-23', 'Turnaround Jump Shot-13',
       'Turnaround Jump Shot-14', 'Jump Shot-4', 'Turnaround Jump Shot-16',
       'Jump Shot-21', 'Tip Shot-0', 'Running Hook Shot-0',
       'Running Jump Shot-16', 'Running Jump Shot-5',
       'Alley Oop Dunk Shot-0', 'Running Jump Shot-20', 'Dunk 

In [14]:
len(kobe["at_sd"].unique())

514

514 values of categorical variable out of a sample of about 30,000 events might be too many. We shall remember it when using these variables in the prediction model.

## Creating trend variable: percentage of made shots over the previous three months

#### Note: incomplete.

After observing data on Tableau, I'm not sure a trend variable can be relevant for a prediction model. Anyway let's try to create it as an exercise.

We drop the submission set.

In [17]:
kobe_trend = kobe[pd.notnull(kobe["shot_made_flag"])]

We create a new column adding up 'game_year' and 'game_month' labels.

In [18]:
kobe_trend["game_year_month"] = (kobe_trend["game_year"].astype(str)) + "-" + (kobe_trend["game_month"].astype(str))

/usr/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Now we create a dataframe after doing a groupby to find out the number of shots for every year/month.

In [19]:
kobe_trend_totalshots = pd.DataFrame({"total_count" : kobe_trend.groupby(["game_year_month"]).size()}).reset_index()

In [67]:
kobe_trend_totalshots.head(3)

game_year_month  total_count
0         1996-11           46
1         1996-12           23
2          1997-1           88

We repeat the process to calculate the number of shots in for every year/month.

In [20]:
kobe_trend_madeshots = pd.DataFrame({"made_count" : kobe_trend.groupby(["game_year_month", "shot_made_flag"]).size()}).reset_index()

In [21]:
kobe_trend_madeshots = kobe_trend_madeshots[kobe_trend_madeshots["shot_made_flag"]==1]

In [70]:
kobe_trend_madeshots.head(3)

game_year_month  shot_made_flag  made_count
1         1996-11               1          19
3         1996-12               1           7
5          1997-1               1          43

Now we merge the dataframes.

In [22]:
kobe_trend_merged = pd.merge(kobe_trend_madeshots, kobe_trend_totalshots, on="game_year_month")

In [72]:
kobe_trend_merged.head(3)

game_year_month  shot_made_flag  made_count  total_count
0         1996-11               1          19           46
1         1996-12               1           7           23
2          1997-1               1          43           88

We turn the values of 'game_year_month' into datetime values in order to work with time series.

In [73]:
kobe_trend_merged.dtypes

game_year_month     object
shot_made_flag     float64
made_count           int64
total_count          int64
dtype: object

In [23]:
kobe_trend_merged["game_year_month"] = pd.to_datetime(kobe_trend_merged["game_year_month"])

We sort the values in chronological order.

In [24]:
kobe_trend_merged = kobe_trend_merged.sort("game_year_month")

We standarize the day for all dates.

In [25]:
kobe_trend_merged["game_year_month"] = kobe_trend_merged["game_year_month"].apply(lambda dt: dt.replace(day=1))

In [77]:
len(kobe_trend_merged)

140

In [78]:
print kobe_trend_merged["game_year_month"]

0     1996-11-01
1     1996-12-01
2     1997-01-01
6     1997-02-01
7     1997-03-01
8     1997-04-01
9     1997-05-01
3     1997-10-01
4     1997-11-01
5     1997-12-01
10    1998-01-01
11    1998-02-01
12    1998-03-01
13    1998-04-01
14    1998-05-01
16    1999-02-01
17    1999-03-01
18    1999-04-01
19    1999-05-01
15    1999-12-01
20    2000-01-01
24    2000-02-01
25    2000-03-01
26    2000-04-01
27    2000-05-01
28    2000-06-01
21    2000-10-01
22    2000-11-01
23    2000-12-01
29    2001-01-01
         ...    
110   2011-01-01
112   2011-02-01
113   2011-03-01
114   2011-04-01
115   2011-05-01
111   2011-12-01
116   2012-01-01
120   2012-02-01
121   2012-03-01
122   2012-04-01
123   2012-05-01
117   2012-10-01
118   2012-11-01
119   2012-12-01
124   2013-01-01
126   2013-02-01
127   2013-03-01
128   2013-04-01
125   2013-12-01
129   2014-10-01
130   2014-11-01
131   2014-12-01
132   2015-01-01
133   2015-10-01
134   2015-11-01
135   2015-12-01
136   2016-01-01
137   2016-02-

At this point I have achieved to create a function that returns a list of lists with the three previous dates if they appear in 'game_year_month' of kobe_trend_merged db.

In [26]:
def get_previous3months(df):
    p3m_ls = df["game_year_month"].apply(lambda date: [date - pd.DateOffset(months=1), date - pd.DateOffset(months=2), date - pd.DateOffset(months=3)])
    return p3m_ls

In [27]:
previous3months_lists = get_previous3months(kobe_trend_merged)

In [28]:
previous3months_lists

0      [1996-10-01 00:00:00, 1996-09-01 00:00:00, 199...
1      [1996-11-01 00:00:00, 1996-10-01 00:00:00, 199...
2      [1996-12-01 00:00:00, 1996-11-01 00:00:00, 199...
6      [1997-01-01 00:00:00, 1996-12-01 00:00:00, 199...
7      [1997-02-01 00:00:00, 1997-01-01 00:00:00, 199...
8      [1997-03-01 00:00:00, 1997-02-01 00:00:00, 199...
9      [1997-04-01 00:00:00, 1997-03-01 00:00:00, 199...
3      [1997-09-01 00:00:00, 1997-08-01 00:00:00, 199...
4      [1997-10-01 00:00:00, 1997-09-01 00:00:00, 199...
5      [1997-11-01 00:00:00, 1997-10-01 00:00:00, 199...
10     [1997-12-01 00:00:00, 1997-11-01 00:00:00, 199...
11     [1998-01-01 00:00:00, 1997-12-01 00:00:00, 199...
12     [1998-02-01 00:00:00, 1998-01-01 00:00:00, 199...
13     [1998-03-01 00:00:00, 1998-02-01 00:00:00, 199...
14     [1998-04-01 00:00:00, 1998-03-01 00:00:00, 199...
16     [1999-01-01 00:00:00, 1998-12-01 00:00:00, 199...
17     [1999-02-01 00:00:00, 1999-01-01 00:00:00, 199...
18     [1999-03-01 00:00:00, 19

By manually selecting the row of 'kobe_trend_merged' I extract the chart with the values of the previous three months (if there is any).

In [29]:
previous3months_table = kobe_trend_merged[kobe_trend_merged["game_year_month"].isin(previous3months_lists[24])]

In [30]:
previous3months_table

game_year_month  shot_made_flag  made_count  total_count
15      1999-12-01               1         102          218
20      2000-01-01               1          93          216

Next I calculate the accuracy rate (the sought value).

In [31]:
previous3months_table[["made_count", "total_count"]] = previous3months_table[["made_count", "total_count"]].astype(float)
accuracy_prev3m = 1000 * np.around(sum(previous3months_table["made_count"]) / sum(previous3months_table["total_count"]), 3)
accuracy_prev3m

/usr/lib64/python2.7/site-packages/pandas/core/frame.py:2148: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


449.0

I imagine I should add a for comprehension function to extract in a loop the data contained in the table and create an if-else condition to asign 0.0 value to the dates of 'kobe_trend_merged['game_year_month']' with no sample of shots during the previous three months. From here it should be easy to calculate the accuracy rate.

_______________________________________________________________________________________________________________________

We export the dataset with the new variables to use it in the next step.

In [81]:
kobe.to_csv("./data/kobe_new_variables.csv", sep= ",")